### ISSUE TITLE
LG-234 mpMetrics-2.2 implementation

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#8653]
AHA: https://bigblue.aha.io/features/LG-234
ID: https://github.ibm.com/was-liberty/liberty-docs/issues/1648

Target: 19.0.0.12
Feature completion: 11/16/2019

**Comment by Yushan-Lin** [comment]
Initial mpMetrics-2.1 Implementation: https://github.com/OpenLiberty/open-liberty/issues/9068

**Comment by marikaj123** [comment]
If the feature is planned to GM for the next release, please add the label target:ga.

**Comment by jhanders34** [comment]
Joe did performance comparison of mpMetrics 2.1 and 2.0 and it showed performance to be better as expected due to the API updates. Giving perf approval.

**Comment by marikaj123** [comment]
When ready, please add the label WAD or No WAD approval, as well as, add the label target:ga in order to become visible to the focal point owners

**Comment by Channyboy** [comment]
@gscottj This feature does not add any new or change any existing UI.

**Comment by sabolo** [comment]
In reviewing the FAT I have found that defect 268309 was not in fact fixed.  I have raised a follow on defect 268458 to get this product issue addressed.  This will need to be resolved prior to receiving approval for the FAT work.

**Comment by skasund** [comment]
@Yushan-Lin   As long as this is documented in the KC for mpMetrics-2.1, we don't need another STE.  Please let me know if it'll be documented, and then I can provide the STE approval.  Thanks.

**Comment by Yushan-Lin** [comment]
@skasund The changes will be documented in KC

**Comment by skasund** [comment]
@Yushan-Lin  Thanks.  I've approved this feature.

**Comment by Yushan-Lin** [comment]
@donbourne Serviceability Approval Comment - Please answer the following questions for serviceability approval:

    WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

    Yes, the WAD identifies the migration problem when users compile their application with Microprofile Metrics 2.0 API and running with mpMetrics-2.1 feature, there will be compilation error. Thus, users must compile their application with Microprofile Metrics 2.0 API to run with mpMetrics-2.1 feature. Error messages will be logged in the messages.log. No major functional change was made for this mpMetricsh-2.1 feature, as everything is similar to mpMetrics-2.0 feature.

    Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
    a) What problem paths were tested and demonstrated?
    b) Who did you demo to?
    c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

    For mpMetrics-2.1,  The common problem paths are similar to mpMetrics-2.0, and was previously tested and demo'ed to SVT.

    SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
    a) Who conducted SVT tests for this feature?

    No SVT was done for this feature, as it only had minor changes compared to mpMetrics-2.0.

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?

    N/A

    Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.

    WAS L3: Liberty Log Analytics

**Comment by Yushan-Lin** [comment]
@marikaj123 Translation is not needed, may I have globalization approval?
@sabolo May I have FAT approval? #9323
@chirp1 May I have ID approval? https://github.ibm.com/was-liberty/liberty-docs/issues/1648

**Comment by chirp1** [comment]
Discussed docs with Don/Yushan. OL updates will be prioritized on OL docs backlog. Might initially need a blog post with extra detail if doc updates are not addressed by 190012. Approving.

**Comment by sabolo** [comment]
I am looking to close out the FAT review however I see no evidence of the tck bucket running in the latest cross platform run (com.ibm.ws.microprofile.metrics.2.2_fat_tck).  There is evidence of this running a few weeks ago.  Investigation is needed to identify why we are no longer seeing results.

**Comment by sabolo** [comment]
There is a possibility this is not caused by issues with this bucket and instead by another one killing the testing... This is being looked at as tests are running successfully in the release builds.

**Comment by sabolo** [comment]
I am seeing the tests running in the release build but not in the full fat runs (which the cross platform testing also uses).  Investigation needed.

**Comment by Channyboy** [comment]
Hi @sabolo, the `mpMetrics-2.2` feature and tests were promoted to master yesterday so I don't think they've ran a full SOE yet.

However, the changes from `2.1` to `2.2` were to remove some changes added into `2.1`.
In affect, there's nothing _new_ being added in `2.2`.

**Comment by sabolo** [comment]
Please can you clarify further what this feature covers.  This was originally the Metrics 2.1 implementation and is not 2.2.  You comment says you are removing things from 2.1 however that was all part of this release.

Your feature test summary only mentions 2.2 testing being delivered which I don't believe is correct.  At this time I am not clear about the full scope of this feature, what testing was delivered and what we should be expecting to be running.  Please can you expand on the feature test summary and clearly detail the work being done here.

**Comment by Channyboy** [comment]
@sabolo 
The TCK of `MicroProfile Metrics 2.1` was expanded on top of the TCK of `MicroProfile Metrics 2.0`.
There was one change that affected the code to be changed in an implementation of `MicroProfile Metrics 2.1`.
Similarily, the metrics community modified the TCK to test for this change.

It was decided that this change in `MicroProfile Metrics 2.1` was a breaking change and that a  `MicroProfile Metrics 2.2` release  was needed to revert the change into `MicroProfile Metrics  2.1`.
In effect the code for `MicroProfile Metrics  2.2` is the same as `2.0`. The TCK had to be reverted too and similarily the TCK for `2.2` tests the same coverage as `2.0`.

**Comment by Channyboy** [comment]
Epic Summary:

The epic for LG-234 was to cover the implementation of `MicroProfile Metrics 2.1`.
However, as there was a breaking change in the `2.1` release a `2.2` release was needed to revert the breaking change. As this feature/epic was to cover the delivery and implementation of `Microprofile Metrics 2.1` it is now coverted to track the delivery and implementation of `MicroProfile metrics 2.2`.

As a result, _since_ the release/implementation of the `MicroProfile Metrics 2.0` there is no net change in functionality from `Microprofile Metrics 2.2`.

**Comment by marikaj123** [comment]
Thank you for confirming the feature did make the GM release 190012. When you can please close the feature.  Thank you.
